src/github.com/matrix-org/dendrite/clientapi/readers/profile.go

```diff
@@ -175,13 +238,96 @@ func SetDisplayName(
 	}
 }
 
-func getLocalPart(userID string) string {
-	if !strings.HasPrefix(userID, "@") {
-		panic(fmt.Errorf("Invalid user ID"))
+func buildMembershipEvents(
+	memberships []authtypes.Membership, db *accounts.Database,
+	oldProfile *authtypes.Profile, changedKey string, newValue string,
+	userID string, cfg *config.Dendrite, queryAPI api.RoomserverQueryAPI,
+) ([]gomatrixserverlib.Event, error) {
+	ev := []gomatrixserverlib.Event{}
+
+	for _, membership := range memberships {
+		prevContent := events.MemberContent{
+			Membership:  "join",
+			DisplayName: oldProfile.DisplayName,
+			AvatarURL:   oldProfile.AvatarURL,
+		}
+
+		prev := prevMembership{
+			UserID:      userID,
+			PrevID:      membership.EventID,
+			PrevContent: prevContent,
+		}
+
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   membership.RoomID,
+			Type:     "m.room.member",
+			StateKey: &userID,
+		}
+
+		if err := builder.SetUnsigned(prev); err != nil {
+			return nil, err
+		}
+
+		content := events.MemberContent{
+			Membership: "join",
+		}
+
+		if changedKey == "displayname" {
+			content.DisplayName = newValue
+			content.AvatarURL = oldProfile.AvatarURL
+		} else if changedKey == "avatar_url" {
+			content.DisplayName = oldProfile.DisplayName
+			content.AvatarURL = newValue
+		}
```

Comment:
This might be simpler if this function was passed the new profile as a authtypes.Profile